src/shared/sam/cli/samCliCommand.ts

```diff
@@ -0,0 +1,94 @@
+/*!
```

Comment:
I feel like this leaks too much of the underlying implementation. Callers shouldn't have to know about or interact with the underlying process. What do you think of something like this?

```typescript
abstract class SamCliCommand<T> {
    ...

    protected readonly samCliLocation: string | undefined

    protected constructor(protected readonly config: SamCliConfiguration) {
        this.samCliLocation = config.getSamCliLocation()
    }

    public abstract execute(): Thenable<T>
}

class SamCliInfoCommand extends SamCliCommand<SamCliInfoReponse> {
    ...

    public constructor(config: SamCliConfiguration = new DefaultSamCliConfiguration(
        new DefaultSettingsConfiguration(extensionSettingsPrefix),
        new DefaultSamCliLocationProvider()
    )) {
        super(config)
    }

    public async execute(): Promise<SamCliInfoReponse>() {
        // start the process, await its completion, and return the output (or throw an error)
    }
}
```

Reply:
I figured each of the sam commands would fall under child_process use, vscode.Task use, or both. I'd set it up to get reusability out of the child_process and vscode.Task portions, but currently prefer your alternative. Going to try it.